### ISSUE TITLE
Make lost devices appear to function as much as possible

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#1629]
From a discussion with @jdarpinian.

To reduce the fragility of the API, lost devices should appear to function as usual, as much as possible. In particular, we should avoid causing exceptions or promise rejections due to lost devices, as this could cause unexpected interruptions in arbitrary places in user code, potentially preventing them from recovering correctly.

- [x] mapAsync #3937
- [x] onSubmittedWorkDone #^
- [x] popErrorScope #4051
- [x] uncapturederror #4017
- [x] createRenderPipelineAsync/createComputePipelineAsync #3296
- [ ] ... and #4752
- [ ] getCompilationInfo #^
- [x] spec should suggest an unconditional console warning on device loss (since it's rare), so that the fact that most things continue to work without error doesn't confuse developers #4052

**Comment by jdarpinian** [comment]
Out of curiosity I went and tested a bunch of WebGL sites randomly sampled from the "Cool WebGL Stuff" section of the WebGL WG meeting notes. I found that approximately half of them support restoring a lost context; the rest don't try. Hopefully we can make it easier in WebGPU to improve that percentage.

**Comment by jdarpinian** [comment]
The spec seems to be mostly designed for this already. The only obvious thing I saw in skimming the spec was that popErrorScope is specced to reject its promise if the device is lost, which should be easy to change. But I think it's likely there will be other issues discovered by API users.

**Comment by kainino0x** [comment]
FTR: I also want to make mapAsync (and onSubmittedWorkDone) still succeed even if the device is lost. I think this is actually the one that applications will run into most often.

**Comment by kainino0x** [comment]
We could alternatively leave those pending forever rather than rejecting or resolving them. Will need to consider the design.

**Comment by greggman** [comment]
Is pending forever problematic with GC/cleanup? If I've closed over a bunch of stuff and my promise is never fulfilled. Example:

```
{
   const buf = new Float32Array(1000000);
   await someFunctionThatNeverResolvesNorRejects();
   doSomeCleanup();
}
```

`buf` is never reclaimed because it appears still in use and `doSomeCleanup` is never called.

**Comment by kainino0x** [comment]
It shouldn't cause problems with GC - you could just as easily do `await new Promise(() => {})`. My understanding is the GC  sees it as the promise resolver holding a reference to the continuation. If the promise resolver becomes unreachable, the continuation becomes unreachable, and then `buf` should become unreachable and get GCed.

**Comment by greggman** [comment]
> you could just as easily do await new Promise(() => {})

That's the user choosing the never return, not the system out of the user's control

> My understanding is the GC sees it as the promise resolver holding a reference to the continuation. If the promise resolver becomes unreachable, the continuation becomes unreachable, and then buf should become unreachable and get GCed.

Maybe? if the spec is specific about what "leave those pending forever means"

But it does still cause the other issue, `doSomeCleanup` is never called. It feels hard to reason about if you expect a promise that you somehow have to track things that might need to be cleaned up beacuse the promise may neither reject nor resolve

```
const things = allocateSomethings();
try {
   await someFunctionThatNeverResolvesNorRejects();
} finally {
   cleanup(things);
}
```

**Comment by kainino0x** [comment]
I agree with the problem of doSomeCleanup() not getting called, I should have said that.

The problem with rejecting is that only works if users explicitly handle the possibility of random device loss (e.g. by using a try-finally) *every time* they use async WebGPU functions. Realistically, no one is going to do that.

So I think resolving the map (even though it's a wasted allocation) is probably the best thing to do. Fortunately it only happens in a pretty exceptional case.

> That's the user choosing the never return, not the system out of the user's control

My point was just, would the GC leak the Float32Array if you did this? It shouldn't, because it's perfectly valid user code that shouldn't break GC.

```js
{
   const buf = new Float32Array(1000000);
   await new Promise(() => {});
   doSomeCleanup();
}
```

The promise in question is returned by our API, but it's still just a promise.

**Comment by greggman** [comment]
I think I'm still confused by how GC works here

```
{
   const buf = new Float32Array(1000000);
   await new Promise(() => {});
   doSomeCleanup();
}
```

The example above, maybe? But knowing whether any arbitrary promise is going to resolve or not is the halting problem. So when it's not a no-op as above, how does GC know the promise will never resolve?

I basically have

```
{
   const buf = new Float32Array(1000000);
   await new Promise(resolve => callIntoWebGPUWithCompletionCallback(resolve));
   doSomeCleanup(buf);
}
```

JS's `await` is holding a refernce the promise. You're saying somehow WebGPU tells the promise that `resolve` is unreferenced and so the promise knows it will never be resolved and it can GC itself, and some how magically this scope in the code just exits, gets cleaned, buf is freed, up and `doSomeCleanup` is never called?

That seems really hard to reason about. That basically any promise in WebGPU may never resolve nor reject if the device is lost.  PS: I'm not suggesting they should reject, maybe they should resolve, but never do either seems like it has issues.

**Comment by kainino0x** [comment]
A promise can only resolve if its resolver function still exists. It can exist either in JS or in the browser's native code, but it must exist somewhere, and the GC is aware of whether it does.

In your example `await` isn't holding a reference to the promise, it's the other way around: the promise resolve holds the promise, which holds the .then callback, which holds the continuation to continue executing after the await. It's clearer if you write it without async syntax:

```
const buf = new Float32Array(1000000);
const promise = new Promise(resolve => callIntoWebGPUWithCompletionCallback(resolve));
promise.then(() => {
    doSomeCleanup(buf);
});
```

or do the equivalent thing without promises at all:

```
const buf = new Float32Array(1000000);
callIntoWebGPUWithCompletionCallback(function myCallback() {
    doSomeCleanup(buf);
});
```

In this case the browser refs `myCallback` which refs (by closure) `buf`. Then if the browser drops myCallback `buf` becomes unreachable.

**Comment by kainino0x** [comment]
I should also mention that Promises are just a standard library thing. They existed as a design pattern before they were added to the language, and can be written in plain ES5. So in principle they can't break GC any more than any other plain ES5 code can.

**Comment by greggman** [comment]
I'm not sure what your point is. My point is I feel like Promises need to resolve or reject always, even on device.destroy and device lost

**Comment by kainino0x** [comment]
My point is just that Promises don't fundamentally have to ever resolve/reject, and if they don't it doesn't break GC. That's not to say it's the correct choice here, but GC doesn't force the issue.

**Comment by litherum** [comment]
Today, editors agreed that mapAsync should reject if the device is lost (for v1)

And editors don't know what to do about popErrorScope (for now, needs more discussion).

**Comment by kainino0x** [comment]
Editors discussed today what to do with popErrorScope(). The following options were considered:

- Simply don't resolve the promise.
- Reject the promise (the current spec behavior).

Both not good because they interrupt the normal flow of the application. Applications should be able to assume device loss (which can happen at any time) won't inject interruptions at random points in their code.

- Report an 'internal' WebGPU error.

We expect an application may be watching for WebGPU errors when it expects to react to it in some way. For 'internal' errors, for example, they may expect to recompile with a simpler compute pipeline. We do not want to trigger this behavior, because doing so is not useful. (Same for 'out-of-memory'.)

- Generate a 'validation' WebGPU error if you do any operation.

'validation' errors are somewhat different, in that we don't generally expect any kind of meaningful recovery from them - only perhaps showing an error message to the user (e.g. on a ShaderToy type website).

This error should never reach the 'uncapturederror' handler, because this would provide a redundant (and unreliable) signal: we prefer to signal problems in only one way so that developers never have to worry about which signal is the correct one to pay attention to (or if they have to handle both).

It would be peculiar, though not out of the question, to generate an error that can never reach the 'uncapturederror' handler.
**Additional post-meeting observation:** Actually not a problem if 'uncapturederror' simply never fires after device loss, which is probably ideal (and we should spec).

- Return a 'validation' error from popErrorScope() with no regard for the contents of the scope.

This makes slightly more sense since it reduces the amount of tracking that needs to happen on a lost device. However, it still provides a redundant signal (device.lost, vs validation error coming from error scopes in the code).

- Pretend everything is fine, always return `null`.

Avoids the more concrete problems described above.
**Additional post-meeting observation:** Could be slightly confusing for developers, but this can be solved with clear console warnings.

**This is the Editors' recommendation.**

**Comment by kainino0x** [comment]
In the next editors' meeting, we should briefly discuss the 2 observations I added above.

We should also briefly discuss create\*PipelineAsync. I'm pretty sure this has been discussed somewhere at some point, so should dig through minutes for #3296. IIRC we said it was fine if it just pretends to work (promise doesn't reject even if you provide a bogus descriptor), but I want to verify that.

**Comment by kainino0x** [comment]
Editors' meeting:

Neither of the observations above changed our recommendations. It would be good to make a spec note about the console warnings. I filed #4017 about the uncapturederror thing.

We also think create\*PipelineAsync should pretend to work to avoid disrupting the flow of the application: applications will definitely sometimes use async pipeline creation without error handling, and `await` would throw an exception and cause a disruption at an unpredictable point in the code.

**Comment by kainino0x** [comment]
Done!

**Comment by kainino0x** [comment]
Reopening because I think a careful reading of createRenderPipelineAsync/createComputePipelineAsync will actually reject the promise with "validation" error:

> invalid due to an validation error
Reject promise with a GPUPipelineError with reason "validation".

> If any of the following conditions are unsatisfied: generate a validation error, make pipeline invalid, and stop.
layout is `valid to use with` this.

> ... `valid to use with`...:
> object.[[device]] must be valid.

This may not be the only instance of the spec being incorrect due to that particular line of `valid to use with`.

**Comment by kainino0x** [comment]
Once the above is fixed, createPipelineAsync never fails after device loss. But it's also unclear whether there can be a race between createPipelineAsync completion and device loss. Should be clarified. I would probably write it such that, if there is a race between device loss (destruction or natural loss) and async pipeline creation:
- if lost resolves first, then createPipelineAsync will succeed
- if createPipelineAsync resolves first, then it will report its validation result.

**Comment by kainino0x** [comment]
Editors meeting: The final device loss check should be in the device timeline, before it sends the message to the content timeline about whether to resolve or reject. The current device timeline checks need to be split in half so there are steps before and after the pipeline creation actually occurs.

This will also allow pipeline creations to complete out of order as intended (by the section that describes promise ordering)

**Comment by Kangz** [comment]
I'm noticing failures with `webgpu:api,validation,state,device_lost,destroy:createRenderPipelineAsync` that say "Invalid async creation should only fail if the device is not yet lost." but it seems that the spec doesn't require this yet. Did the CTS land before the editorial change?

**Comment by kainino0x** [comment]
createComputePipelineAsync, createRenderPipelineAsync, and getCompilationInfo will finally be fixed with #4752.
